File: scripts/persistence-probe.mjs
// SPDX-License-Identifier: Apache-2.0
// Synthetic subprocess peer for file interchange, crash recovery and races.
import {readFileSync,writeFileSync,existsSync} from 'node:fs';
import {DatabaseSync} from 'node:sqlite';
import {WorkflowStore,StoreError} from '../implementations/typescript/packages/api-server/dist/persistence.js';
import {SqliteBackend} from '../implementations/typescript/packages/api-server/dist/sqlite.js';
import {runFixture,secret} from './persistence-fixtures.mjs';
const path=process.argv[2], options=JSON.parse(readFileSync(0,'utf8'));
if(options.mode==='fixture') {
  console.log(JSON.stringify(await runFixture(path)));
} else if(options.mode==='crashUncommitted') {
  const db=new DatabaseSync(path);
  db.exec('PRAGMA cache_size=1; PRAGMA cache_spill=ON; BEGIN IMMEDIATE;');
  db.prepare('UPDATE psp_records SET body=?').run(JSON.stringify({uncommitted:'x'.repeat(65536)}));
  process.exit(0);
} else {
  const backend=new SqliteBackend(path,options.epoch??'peer-epoch',()=>options.now??100);
  const store=new WorkflowStore(backend,{resumeSecret:secret,authorizePersistence:async()=>{
    if(options.barrier) {
      writeFileSync(options.barrier.ready,'ready');
      const deadline=Date.now()+15000;
      while(!existsSync(options.barrier.go)) {
        if(Date.now()>deadline) throw Error('barrier timeout');
        await new Promise(resolve=>setTimeout(resolve,20));
      }
    }
    return true;
  }});
  let result;
  try {result={result:await store.execute(options.actor,options.command)};}
  catch(e) {if(!(e instanceof StoreError)) throw e;result={error:e.code};}
  if(options.crashAfterCommit) process.stdout.write(JSON.stringify(result)+'\n',()=>process.exit(0));
  else {backend.close();console.log(JSON.stringify(result));}
}

File: scripts/persistence_probe.py
# SPDX-License-Identifier: Apache-2.0
import json
import os
from pathlib import Path
import sqlite3
import sys
import time
from persistence_fixtures import SECRET, run_fixture
from psp_cdl_api_server.persistence import WorkflowStore, StoreError
from psp_cdl_api_server.sqlite import SqliteBackend

sys.stdin.reconfigure(encoding="utf-8")
sys.stdout.reconfigure(encoding="utf-8")
path, options = sys.argv[1], json.load(sys.stdin)
if options.get("mode") == "fixture":
    print(json.dumps(run_fixture(path)))
elif options.get("mode") == "crashUncommitted":
    db = sqlite3.connect(path, isolation_level=None)
    db.execute("PRAGMA cache_size=1")
    db.execute("PRAGMA cache_spill=ON")
    db.execute("BEGIN IMMEDIATE")
    db.execute("UPDATE psp_records SET body=?", (json.dumps({"uncommitted": "x" * 65536}),))
    os._exit(0)
else:
    backend = SqliteBackend(path, options.get("epoch", "peer-epoch"), lambda: options.get("now", 100))
    def permit(*_):
        if barrier := options.get("barrier"):
            Path(barrier["ready"]).write_text("ready", encoding="utf-8")
            deadline = time.monotonic() + 15
            while not Path(barrier["go"]).exists():
                if time.monotonic() > deadline:
                    raise RuntimeError("barrier timeout")
                time.sleep(0.02)
        return True
    store = WorkflowStore(backend, resume_secret=SECRET, authorize_persistence=permit)
    try:
        result = {"result": store.execute(options["actor"], options["command"])}
    except StoreError as exc:
        result = {"error": exc.code}
    if options.get("crashAfterCommit"):
        print(json.dumps(result, ensure_ascii=False), flush=True)
        os._exit(0)
    backend.close()
    print(json.dumps(result, ensure_ascii=False))
